## Project Title:
E Commerce Funnel Analysis 

Understanding Customer Drop-off and Conversion Optimisation

## Background

In e-commerce, understanding how customers move through 
the purchase journey is critical for business growth. 
This analysis explores customer behaviour from product 
view to purchase, identifying where potential buyers 
drop off and providing recommendations to improve 
conversion rates.

## Business Problem 

An e-commerce business wants to understand where 
customers are dropping off in the purchase journey 
and what can be done to improve conversion rates.

## Aims and Objectives 


**Aim:**
To analyse customer behaviour across the e-commerce 
purchase funnel, identify drop-off points and provide 
actionable recommendations to improve conversion rates 
and revenue.

**Objectives:**
1. Map the full customer journey from product view 
   to purchase
2. Calculate conversion rates at each funnel stage
3. Identify where the highest drop-offs occur
4. Segment the funnel by product category and brand
5. Provide data-driven recommendations to reduce 
   drop-off and improve conversion

## Business problem 
An e-commerce business wants to understand where 
customers are dropping off in the purchase journey 
and what can be done to improve conversion rates.
This business has identified a gap in understanding 
how customers move through the purchase journey. 
Specifically, it is unclear where potential buyers 
are dropping off, which product categories experience 
the highest abandonment and what factors influence 
whether a customer completes a purchase or leaves.

Without this visibility, marketing spend, product 
placement and customer experience decisions remain 
reactive rather than data-driven.

## Import  Library 

In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 
import matplotlib.pyplot as plt


## Import E Commerce Dataset 

In [2]:
df = pd.read_csv(r"C:\Users\computer1\Desktop\ecommerce-funnel-analysis\events.csv")
df

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2020-09-24 11:57:06 UTC,view,1996170,2144415922528452715,electronics.telephone,NaN,31.90,1515915625519388267,LJuJVLEjPT
1,2020-09-24 11:57:26 UTC,view,139905,2144415926932472027,computers.components.cooler,zalman,17.16,1515915625519380411,tdicluNnRY
2,2020-09-24 11:57:27 UTC,view,215454,2144415927158964449,NaN,NaN,9.81,1515915625513238515,4TMArHtXQy
3,2020-09-24 11:57:33 UTC,view,635807,2144415923107266682,computers.peripherals.printer,pantum,113.81,1515915625519014356,aGFYrNgC08
4,2020-09-24 11:57:36 UTC,view,3658723,2144415921169498184,NaN,cameronsino,15.87,1515915625510743344,aa4mmk0kwQ
...,...,...,...,...,...,...,...,...,...
885124,2021-02-28 23:55:01 UTC,view,953226,2144415927553229037,NaN,NaN,219.94,1515915625611023730,FRLqIttxKU
885125,2021-02-28 23:58:05 UTC,view,1715907,2144415927049912542,electronics.video.tv,starwind,80.03,1515915625611024014,g6WqPf50Ma
885126,2021-02-28 23:58:09 UTC,view,4170534,2144415939364389423,electronics.clocks,amazfit,64.92,1515915625611024020,xNIJBqZdkd
885127,2021-02-28 23:58:14 UTC,view,888273,2144415921932861531,electronics.telephone,NaN,10.16,1515915625611024030,9pCbKMIcSx


In [3]:
df.head()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2020-09-24 11:57:06 UTC,view,1996170,2144415922528452715,electronics.telephone,NaN,31.90,1515915625519388267,LJuJVLEjPT
1,2020-09-24 11:57:26 UTC,view,139905,2144415926932472027,computers.components.cooler,zalman,17.16,1515915625519380411,tdicluNnRY
2,2020-09-24 11:57:27 UTC,view,215454,2144415927158964449,NaN,NaN,9.81,1515915625513238515,4TMArHtXQy
3,2020-09-24 11:57:33 UTC,view,635807,2144415923107266682,computers.peripherals.printer,pantum,113.81,1515915625519014356,aGFYrNgC08
4,2020-09-24 11:57:36 UTC,view,3658723,2144415921169498184,NaN,cameronsino,15.87,1515915625510743344,aa4mmk0kwQ


## Data Understanding 

#### In this section we explore the structure of the dataset,examine data types, duplicates,  identify missing values and understandthe distribution of key variables before proceeding to cleaning.

### Check for duplicates 

In [4]:
df.duplicated().sum()

np.int64(655)

#### Check for missing values 

In [5]:
df.isnull().sum()

event_time            0
event_type            0
product_id            0
category_id           0
category_code    236219
brand            212364
price                 0
user_id               0
user_session        165
dtype: int64

#### Check for summary statistics 

In [6]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
product_id,885129.0,1.906621e+06,1.458708e+06,1.020000e+02,6.988030e+05,1.452883e+06,3.721194e+06,4.183880e+06
category_id,885129.0,2.144423e+18,6.165105e+14,2.144416e+18,2.144416e+18,2.144416e+18,2.144416e+18,2.227847e+18
price,885129.0,1.463287e+02,2.968077e+02,2.200000e-01,2.646000e+01,6.571000e+01,1.904900e+02,6.477106e+04
user_id,885129.0,1.515916e+18,3.554165e+07,1.515916e+18,1.515916e+18,1.515916e+18,1.515916e+18,1.515916e+18


In [7]:
print("Dataset Shape:", df.shape)
print("\nData Types:")
print(df.dtypes)
print("\nMissing Values:")

Dataset Shape: (885129, 9)

Data Types:
event_time        object
event_type        object
product_id         int64
category_id        int64
category_code     object
brand             object
price            float64
user_id            int64
user_session      object
dtype: object

Missing Values:


#### Observations
- Dataset contains 885,129 rows and 9 columns
- event_time needs converting from object to datetime
- All other data types look appropriate
- Duplicates rows: 652

In [8]:
print(df['event_time'].dtype)
print(df['event_type'].value_counts())

object
event_type
view        793748
cart         54035
purchase     37346
Name: count, dtype: int64


#### Insight:  Three columns contain missing values:
- category_code: 236,219 missing (26.7%)
- brand: 212,364 missing (24.0%)
- user_session: 165 missing (0.02%)

category_code and brand are filled with 'Uncategorised'
since the high percentage makes dropping impractical.
user_session rows are dropped as they represent only
0.02% of the dataset and session ID is critical for
funnel tracking.

## Data Cleaning
- Fix event_time to datetime
- Handle missing values
- Check and drop duplicates
- Save cleaned dataset

#### Convert event_time to datetime

In [9]:
df['event_time'] = pd.to_datetime(df['event_time'])
print(f"event_time dtype: {df['event_time'].dtype}")

event_time dtype: datetime64[ns, UTC]


#### Handle missing values 

In [10]:
df = df.dropna(subset=['user_session'])
df = df.copy()
df['category_code'] = df['category_code'].fillna('Uncategorised')
df['brand'] = df['brand'].fillna('Uncategorised')

print(df.isnull().sum())


event_time       0
event_type       0
product_id       0
category_id      0
category_code    0
brand            0
price            0
user_id          0
user_session     0
dtype: int64


#### Drop duplicates 

In [11]:
df.drop_duplicates()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2020-09-24 11:57:06+00:00,view,1996170,2144415922528452715,electronics.telephone,Uncategorised,31.90,1515915625519388267,LJuJVLEjPT
1,2020-09-24 11:57:26+00:00,view,139905,2144415926932472027,computers.components.cooler,zalman,17.16,1515915625519380411,tdicluNnRY
2,2020-09-24 11:57:27+00:00,view,215454,2144415927158964449,Uncategorised,Uncategorised,9.81,1515915625513238515,4TMArHtXQy
3,2020-09-24 11:57:33+00:00,view,635807,2144415923107266682,computers.peripherals.printer,pantum,113.81,1515915625519014356,aGFYrNgC08
4,2020-09-24 11:57:36+00:00,view,3658723,2144415921169498184,Uncategorised,cameronsino,15.87,1515915625510743344,aa4mmk0kwQ
...,...,...,...,...,...,...,...,...,...
885124,2021-02-28 23:55:01+00:00,view,953226,2144415927553229037,Uncategorised,Uncategorised,219.94,1515915625611023730,FRLqIttxKU
885125,2021-02-28 23:58:05+00:00,view,1715907,2144415927049912542,electronics.video.tv,starwind,80.03,1515915625611024014,g6WqPf50Ma
885126,2021-02-28 23:58:09+00:00,view,4170534,2144415939364389423,electronics.clocks,amazfit,64.92,1515915625611024020,xNIJBqZdkd
885127,2021-02-28 23:58:14+00:00,view,888273,2144415921932861531,electronics.telephone,Uncategorised,10.16,1515915625611024030,9pCbKMIcSx


##### Extracting year, month and day for time trends analysis 

In [12]:
df['date'] = df['event_time'].dt.date
df['hour'] = df['event_time'].dt.hour
df['month_name'] = df['event_time'].dt.strftime('%B')
df['day_of_week'] = df['event_time'].dt.strftime('%A')

print(df[['event_time', 'date', 'hour', 'month_name', 'day_of_week']].head())

                 event_time        date  hour month_name day_of_week
0 2020-09-24 11:57:06+00:00  2020-09-24    11  September    Thursday
1 2020-09-24 11:57:26+00:00  2020-09-24    11  September    Thursday
2 2020-09-24 11:57:27+00:00  2020-09-24    11  September    Thursday
3 2020-09-24 11:57:33+00:00  2020-09-24    11  September    Thursday
4 2020-09-24 11:57:36+00:00  2020-09-24    11  September    Thursday


In [13]:
df

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session,date,hour,month_name,day_of_week
0,2020-09-24 11:57:06+00:00,view,1996170,2144415922528452715,electronics.telephone,Uncategorised,31.90,1515915625519388267,LJuJVLEjPT,2020-09-24,11,September,Thursday
1,2020-09-24 11:57:26+00:00,view,139905,2144415926932472027,computers.components.cooler,zalman,17.16,1515915625519380411,tdicluNnRY,2020-09-24,11,September,Thursday
2,2020-09-24 11:57:27+00:00,view,215454,2144415927158964449,Uncategorised,Uncategorised,9.81,1515915625513238515,4TMArHtXQy,2020-09-24,11,September,Thursday
3,2020-09-24 11:57:33+00:00,view,635807,2144415923107266682,computers.peripherals.printer,pantum,113.81,1515915625519014356,aGFYrNgC08,2020-09-24,11,September,Thursday
4,2020-09-24 11:57:36+00:00,view,3658723,2144415921169498184,Uncategorised,cameronsino,15.87,1515915625510743344,aa4mmk0kwQ,2020-09-24,11,September,Thursday
...,...,...,...,...,...,...,...,...,...,...,...,...,...
885124,2021-02-28 23:55:01+00:00,view,953226,2144415927553229037,Uncategorised,Uncategorised,219.94,1515915625611023730,FRLqIttxKU,2021-02-28,23,February,Sunday
885125,2021-02-28 23:58:05+00:00,view,1715907,2144415927049912542,electronics.video.tv,starwind,80.03,1515915625611024014,g6WqPf50Ma,2021-02-28,23,February,Sunday
885126,2021-02-28 23:58:09+00:00,view,4170534,2144415939364389423,electronics.clocks,amazfit,64.92,1515915625611024020,xNIJBqZdkd,2021-02-28,23,February,Sunday
885127,2021-02-28 23:58:14+00:00,view,888273,2144415921932861531,electronics.telephone,Uncategorised,10.16,1515915625611024030,9pCbKMIcSx,2021-02-28,23,February,Sunday


In [14]:
import os
os.makedirs('data', exist_ok=True)
df.to_csv('data/ecommerce_cleaned.csv', index=False)
print("Cleaned dataset saved!")
print(f"Final shape: {df.shape}")

Cleaned dataset saved!
Final shape: (884964, 13)
